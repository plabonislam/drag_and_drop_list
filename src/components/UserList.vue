<script setup>
import { ref } from 'vue'

const users = ref([])
const loading = ref(false)

const fetchUsers = async () => {
  loading.value = true
  try {
    const response = await fetch('https://jsonplaceholder.typicode.com/users')
    if (!response.ok) throw new Error('Failed to fetch')
    const data = await response.json()
    users.value = data.map((user) => ({
      ...user,
      age: user.age ?? 20 // Default age to 20 if missing
    }))
  } catch (error) {
    console.error(error)
  } finally {
    loading.value = false
  }
}
</script>

<template>
  <div>
    <button @click="fetchUsers">Fetch Users</button>
    <p v-if="loading">Loading...</p>
    <ul>
      <li v-for="(user, index) in users" :key="index">
        {{ user.name }} - {{ user.age }} years old
      </li>
    </ul>
  </div>
</template>
