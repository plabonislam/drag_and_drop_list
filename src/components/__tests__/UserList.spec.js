import { mount ,flushPromises } from "@vue/test-utils";
import UserList from "@/components/UserList.vue";
import { describe, it, expect, vi,beforeEach } from 'vitest'
describe("UserList", () => {  

 it('renders properly',()=>{
  const wrapper = mount(UserList)
  expect(wrapper.html()).toMatchSnapshot();
 })

 it('fetching data',async()=>{
  global.fetch = vi.fn(() =>
    Promise.resolve({
      ok: true,
      json: () =>
        Promise.resolve([
          { name: "Alice", age: 25 },
          { name: "Bob" }, // No age, should default to 20
          { name: "Charlie", age: 30 }
        ])
    })
  );
  const wrapper = mount(UserList)
  // Ensure the list is empty initially
  expect(wrapper.find("li").exists()).toBe(false);
  // Click the button
  await wrapper.find("button").trigger("click");
  wrapper.vm.loading = true ;
  // Check if the loading text appears
  // expect(wrapper.find("p").text()).toBe("Loading...");

  // Wait for the users to load
  await flushPromises();

  console.log(wrapper.vm.loading)
  // Check if the correct user data is displayed
  const listItems = wrapper.findAll("li");
  console.log(listItems.length)
  expect(listItems.length).toBe(3);
  expect(listItems[0].text()).toBe("Alice - 25 years old");
  expect(listItems[1].text()).toBe("Bob - 20 years old"); // Default age
  expect(listItems[2].text()).toBe("Charlie - 30 years old");
});


it('updates loading state', async () => {
  const wrapper = mount(UserList)
  

  wrapper.vm.loading = true ;
   expect(wrapper.vm.loading).toBe(true)
  //  console.log(wrapper.find('p').exists().text())
  // // Check if the loading text appears 
 // expect(wrapper.find('p').text()).toContain('Loading...')
})
})