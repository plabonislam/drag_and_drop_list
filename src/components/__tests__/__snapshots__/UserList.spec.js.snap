// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserList > renders properly 1`] = `
"<div><button>Fetch Users</button>
  <!--v-if-->
  <ul></ul>
</div>"
`;
